File: JavaScript/1-Arrays&Hashing/Problems/1-TwoSum/TwoSumPro.js
/**
 * 1. Arrays & Hashing >> Two Sum
 *
 * Given an array of integers, nums, and an integer target, return the indices of the two numbers
 * in the array that add up to the target.
 *
 * You may assume that each input would have exactly one solution, and you may not use the same element twice.
 * You can return the answer in any order.
 *
 * Example 1:
 * Input: nums = [2,7,11,15], target = 9
 * Output: [0,1]
 * Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
 *
 * Example 2:
 * Input: nums = [3,2,4], target = 6
 * Output: [1,2]
 *
 * Example 3:
 * Input: nums = [3,3], target = 6
 * Output: [0,1]
 *
 * Constraints:
 * 2 <= nums.length <= 10^4
 * -10^9 <= nums[i] <= 10^9
 * -10^9 <= target <= 10^9
 */

/**
 * Approach 1: Brute Force
 *
 * Time Complexity: O(n^2)
 * Space Complexity: O(1)
 *
 * Description:
 * This is the most straightforward approach. We iterate through each element in the array
 * and check the sum of it with every other element. If the sum equals the target, we return the indices.
 *
 * Real-world application:
 * This approach can be used in a very basic data validation scenario where you need to check all possible
 * combinations of values to find a match.  For example, in a very simple, non-performance-critical
 * configuration check, you might use this to ensure no two settings conflict.  It's generally avoided
 * in production code due to its inefficiency.
 */
const twoSumBruteForce = (nums, target) => {
    for (let i = 0; i < nums.length; i++) {
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[i] + nums[j] === target) {
                return [i, j];
            }
        }
    }
    // No solution found (though problem statement guarantees one)
    return [];
};

/**
 * Approach 2: Hash Map (Two-Pass)
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *
 * Description:
 * In this approach, we use a hash map (JavaScript object) to store each number from the array and its index.
 * First, we iterate through the array and store each number and its index in the hash map.
 * Then, we iterate through the array again. For each number, we check if the difference
 * (target - number) exists in the hash map. If it does, and the index is different from the current index,
 * we have found our pair.
 *
 * Real-world application:
 * Hash maps are extremely common.  This two-pass approach could be used in a system that
 * processes user requests, where you need to quickly look up related data.  For example,
 * in an e-commerce system, you might store product IDs and their details in a hash map
 * for fast retrieval during the checkout process.  The first pass pre-loads the data,
 * and the second pass uses it for lookups.
 */
const twoSumHashMapTwoPass = (nums, target) => {
    const numMap = {};

    // Store numbers and their indices in the hash map
    for (let i = 0; i < nums.length; i++) {
        numMap[nums[i]] = i;
    }

    // Find the pair
    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];
        if (numMap[complement] !== undefined && numMap[complement] !== i) {
            return [i, numMap[complement]];
        }
    }

    return [];
};

/**
 * Approach 3: Hash Map (One-Pass)
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *
 * Description:
 * This is an optimized version of the hash map approach. Instead of iterating through the array twice,
 * we do it in one pass. For each number, we check if its complement (target - number) is already
 * in the hash map. If it is, we found the pair. If not, we add the number and its index to the hash map.
 *
 * Real-world application:
 * This one-pass hash map is even more efficient.  Consider a real-time data processing
 * scenario, like processing a stream of sensor readings.  You might need to identify
 * pairs of readings that satisfy a certain condition (the 'target').  A one-pass hash map
 * allows you to do this very efficiently as the data arrives, without needing to store the
 * entire dataset first.
 */
const twoSumHashMapOnePass = (nums, target) => {
    const numMap = {};

    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];
        if (numMap[complement] !== undefined) {
            return [numMap[complement], i];
        } else {
            numMap[nums[i]] = i;
        }
    }

    return [];
};

/**
 * Approach 4: Sorting and Two Pointers
 *
 * Time Complexity: O(n log n)  (due to sorting)
 * Space Complexity: O(1) or O(n) depending on the sorting algorithm
 *
 * Description:
 * This approach first sorts the array. Then, we use two pointers, one at the beginning and one at the end
 * of the sorted array. We calculate the sum of the numbers at these two pointers.
 * If the sum is equal to the target, we found the pair. If the sum is less than the target, we move the left
 * pointer to the right. If the sum is greater than the target, we move the right pointer to the left.
 *
 * Real-world application:
 * This approach is useful when the input array needs to be sorted anyway for other purposes.
 * For example, imagine you're working on a system that analyzes financial transactions.  You might
 * sort the transactions by amount, and then use a two-pointer approach to find pairs of transactions
 * that meet a specific condition (e.g., sum to a particular fraud threshold).  The sorting step
 * might be a requirement of the overall financial analysis process.
 *
 * Note:  This approach modifies the original array, and it's more difficult to directly get the original indices.
 */
const twoSumSortedTwoPointers = (nums, target) => {
    // Create a copy of the array to avoid modifying the original input
    const sortedNums = [...nums];
    sortedNums.sort((a, b) => a - b); // Sort the copied array

    let left = 0;
    let right = sortedNums.length - 1;

    while (left < right) {
        const sum = sortedNums[left] + sortedNums[right];
        if (sum === target) {
             // Need to find original indices.
             let index1 = -1;
             let index2 = -1;
             for (let i = 0; i < nums.length; i++){
                if (nums[i] === sortedNums[left]){
                    index1 = i;
                    break; // Important:  Find only the *first* occurrence
                }
             }
             for (let i = 0; i < nums.length; i++){
                if (nums[i] === sortedNums[right] && i !== index1){
                    index2 = i;
                    break;
                }
             }
            return [index1, index2];
        } else if (sum < target) {
            left++;
        } else {
            right--;
        }
    }
    return [];
};


/**
 * Approach 5: Using Map.prototype.forEach()
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *
 * Description:
 * This approach uses the forEach method of the Map object.  It's similar to the one-pass hash map,
 * but uses the Map data structure.
 *
 * Real-world application:
 * Useful in scenarios where you're already working with Map objects, or prefer the explicit
 * forEach iteration.  For example, you might be processing data from an API that returns
 * key-value pairs, and you've stored some of that data in a Map.  This approach integrates
 * well with that data structure.
 */
const twoSumUsingMapForEach = (nums, target) => {
    const numMap = new Map();
    let result = [];

    nums.forEach((num, index) => {
        const complement = target - num;
        if (numMap.has(complement)) {
            result = [numMap.get(complement), index];
        } else {
            numMap.set(num, index);
        }
    });
    return result;
};

/**
 * Approach 6: Using Array.prototype.reduce()
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *
 * Description:
 * This approach uses the reduce method of the Array object.  It's a more functional approach.
 *
 * Real-world application:
 * This approach demonstrates a functional programming style.  It could be used in a data
 * transformation pipeline where you're processing data using functional methods.  For example,
 * in a data analysis script, you might use reduce to accumulate information about the data
 * as you process it.
 */
const twoSumUsingReduce = (nums, target) => {
    const numMap = {};
    let result = [];

    nums.reduce((_, num, index) => { // Use _ to indicate unused accumulator
        const complement = target - num;
        if (numMap[complement] !== undefined) {
            result = [numMap[complement], index];
        } else {
            numMap[num] = index;
        }
        return _; // Not used, but required for reduce
    }, null); // Initial value (not needed here)
    return result;
};



// Example Usage
const nums1 = [2, 7, 11, 15];
const target1 = 9;

const nums2 = [3, 2, 4];
const target2 = 6;

const nums3 = [3, 3];
const target3 = 6;

console.log("Example 1:");
console.log("Brute Force:", twoSumBruteForce(nums1, target1));       // Output: [0, 1]
console.log("Hash Map (Two-Pass):", twoSumHashMapTwoPass(nums1, target1));  // Output: [0, 1]
console.log("Hash Map (One-Pass):", twoSumHashMapOnePass(nums1, target1));  // Output: [0, 1]
console.log("Sorted Two Pointers:", twoSumSortedTwoPointers(nums1, target1));  // Output: [0, 1]
console.log("Using Map.forEach():", twoSumUsingMapForEach(nums1, target1));    // Output: [0, 1]
console.log("Using Array.reduce():", twoSumUsingReduce(nums1, target1));      // Output: [0, 1]

console.log("\nExample 2:");
console.log("Brute Force:", twoSumBruteForce(nums2, target2));       // Output: [1, 2]
console.log("Hash Map (Two-Pass):", twoSumHashMapTwoPass(nums2, target2));  // Output: [1, 2]
console.log("Hash Map (One-Pass):", twoSumHashMapOnePass(nums2, target2));  // Output: [1, 2]
console.log("Sorted Two Pointers:", twoSumSortedTwoPointers(nums2, target2));  // Output: [1, 2]
console.log("Using Map.forEach():", twoSumUsingMapForEach(nums2, target2));    // Output: [1, 2]
console.log("Using Array.reduce():", twoSumUsingReduce(nums2, target2));      // Output: [1, 2]

console.log("\nExample 3:");
console.log("Brute Force:", twoSumBruteForce(nums3, target3));       // Output: [0, 1]
console.log("Hash Map (Two-Pass):", twoSumHashMapTwoPass(nums3, target3));  // Output: [0, 1]
console.log("Hash Map (One-Pass):", twoSumHashMapOnePass(nums3, target3));  // Output: [0, 1]
console.log("Sorted Two Pointers:", twoSumSortedTwoPointers(nums3, target3));  // Output: [0, 1]
console.log("Using Map.forEach():", twoSumUsingMapForEach(nums3, target3));    // Output: [0, 1]
console.log("Using Array.reduce():", twoSumUsingReduce(nums3, target3));      // Output: [0, 1]

File: C++/1-Arrays&Hashing/Problems/1-TwoSum/TwoSumPro.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <stdexcept> // For exception handling

// 1. Brute Force Approach
//   - Simplest approach.
//   - Check every possible pair of numbers.
//   - Time Complexity: O(n^2), where n is the size of the input vector.
//   - Space Complexity: O(1).
//   - Real-world application: Useful for small datasets or when a very simple,
//     easily understandable solution is needed, such as in a simple data analysis
//     script for a small experiment.  It's often used as a baseline for
//     comparing the performance of more optimized algorithms.
std::vector<int> twoSumBruteForce(const std::vector<int>& nums, int target) {
    for (size_t i = 0; i < nums.size(); ++i) {
        for (size_t j = i + 1; j < nums.size(); ++j) {
            if (nums[i] + nums[j] == target) {
                return {static_cast<int>(i), static_cast<int>(j)}; // Cast size_t to int
            }
        }
    }
    return {}; // Return empty vector if no solution is found.  Good practice.
}

// 2. Hash Map Approach
//   - Efficient solution using a hash map.
//   - Store each number and its index in the hash map.
//   - For each number, check if its complement (target - number) exists in the hash map.
//   - Time Complexity: O(n) - average case for hash map operations.
//   - Space Complexity: O(n) - for storing the numbers in the hash map.
//   - Real-world application: This is very common in real-world applications.  For
//     example, it could be used in a web service that needs to quickly look up
//     pairs of values (e.g., finding matching product IDs, or processing financial
//     transactions).  Hash maps are highly optimized in modern libraries.
std::vector<int> twoSumHashMap(const std::vector<int>& nums, int target) {
    std::unordered_map<int, int> numMap;
    for (size_t i = 0; i < nums.size(); ++i) {
        int complement = target - nums[i];
        if (numMap.find(complement) != numMap.end()) {
            return {numMap[complement], static_cast<int>(i)}; // Cast size_t to int
        }
        numMap[nums[i]] = static_cast<int>(i); // Cast size_t to int
    }
    return {}; // Return empty vector if no solution is found.
}

// 3. Sorted Array and Two Pointers
//   - Requires the input array to be sorted.
//   - Uses two pointers, one at the beginning and one at the end of the array.
//   - Move the pointers based on whether the sum of the numbers is less than or greater than the target.
//   - Time Complexity: O(n log n) for sorting + O(n) for the two-pointer traversal, so O(n log n).
//   - Space Complexity: O(1) if the sorting is done in place.  Otherwise, O(n) for a copy.
//   - Real-world application: Useful when the input array is already sorted or when
//     sorting is an acceptable preprocessing step.  For example, this approach
//     could be used in a system that processes sorted data, such as in some
//     types of database operations or in specific numerical algorithms.
std::vector<int> twoSumSortedArray(std::vector<int> nums, int target) {
    // Create a vector of pairs to store the original index with the value.
    std::vector<std::pair<int, int>> numsWithIndices;
    for (size_t i = 0; i < nums.size(); ++i) {
        numsWithIndices.push_back({nums[i], static_cast<int>(i)}); // Store value and original index
    }

    // Sort the vector of pairs based on the number.
    std::sort(numsWithIndices.begin(), numsWithIndices.end());

    int left = 0;
    int right = static_cast<int>(numsWithIndices.size()) - 1; // Cast size_t to int

    while (left < right) {
        int sum = numsWithIndices[left].first + numsWithIndices[right].first;
        if (sum == target) {
            // Return the original indices.
            return {numsWithIndices[left].second, numsWithIndices[right].second};
        } else if (sum < target) {
            ++left;
        } else {
            --right;
        }
    }
    return {};
}

// 4. Using Binary Search (After Sorting)
//   - Sort the array.
//   - For each element, use binary search to find its complement.
//   - Time Complexity: O(n log n) for sorting + O(n log n) for the binary searches, so O(n log n).
//   - Space Complexity: O(1) if sorting is done in place.
//   - Real-world application: This approach is suitable when the array is already sorted
//     or when sorting is a reasonable preprocessing step.  Binary search is
//     efficient, making this approach useful in applications where quick lookups
//     are needed in a sorted dataset.
int binarySearch(const std::vector<int>& nums, int target, int excludeIndex) {
    int left = 0;
    int right = static_cast<int>(nums.size()) - 1; // Cast size_t to int
    while (left <= right) {
        int mid = left + (right - left) / 2; // Prevent potential overflow
        if (mid == excludeIndex) { // Skip the excluded index.
           if (mid + 1 <= right) {
               left = mid + 1;
           } else {
               right = mid -1;
           }
           continue;
        }
        if (nums[mid] == target) {
            return mid;
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return -1;
}

std::vector<int> twoSumBinarySearch(std::vector<int> nums, int target) {
    // Create a vector of pairs to store the original index with the value.
    std::vector<std::pair<int, int>> numsWithIndices;
    for (size_t i = 0; i < nums.size(); ++i) {
        numsWithIndices.push_back({nums[i], static_cast<int>(i)}); // Store value and original index
    }

    // Sort the vector of pairs based on the number.
    std::sort(numsWithIndices.begin(), numsWithIndices.end());
    // Extract just the sorted numbers for binary search.
    std::vector<int> sortedNums;
    for (const auto& pair : numsWithIndices) {
        sortedNums.push_back(pair.first);
    }

    for (size_t i = 0; i < sortedNums.size(); ++i) {
        int complement = target - sortedNums[i];
        int complementIndex = binarySearch(sortedNums, complement, static_cast<int>(i)); // Cast size_t to int
        if (complementIndex != -1) {
             // Find original indices.
            int originalIndex1 = numsWithIndices[i].second;
            int originalIndex2 = numsWithIndices[complementIndex].second;
            if (originalIndex1 != originalIndex2)
                return {originalIndex1, originalIndex2};
        }
    }
    return {};
}

// 5. Optimized Hash Map Approach (Slightly different order)
//   -  Same time complexity as the previous hash map solution, but slightly
//      optimized for cases where the complement is found early in the array.
//   - Time Complexity: O(n)
//   - Space Complexity: O(n)
std::vector<int> twoSumOptimizedHashMap(const std::vector<int>& nums, int target) {
    std::unordered_map<int, int> numMap;
    for (size_t i = 0; i < nums.size(); ++i) {
        int complement = target - nums[i];
        // Check if complement exists *before* adding the current number.
        if (numMap.find(complement) != numMap.end()) {
            return {numMap[complement], static_cast<int>(i)};
        }
        numMap[nums[i]] = static_cast<int>(i);
    }
    return {};
}

int main() {
    std::vector<int> nums = {2, 7, 11, 15};
    int target = 9;

    // Brute Force
    std::cout << "Brute Force: ";
    std::vector<int> resultBruteForce = twoSumBruteForce(nums, target);
    if (!resultBruteForce.empty()) {
        std::cout << resultBruteForce[0] << ", " << resultBruteForce[1] << std::endl;
    } else {
        std::cout << "No solution found" << std::endl;
    }

    // Hash Map
    std::cout << "Hash Map: ";
    std::vector<int> resultHashMap = twoSumHashMap(nums, target);
    if (!resultHashMap.empty()) {
        std::cout << resultHashMap[0] << ", " << resultHashMap[1] << std::endl;
    } else {
        std::cout << "No solution found" << std::endl;
    }

    // Sorted Array and Two Pointers
    std::cout << "Sorted Array: ";
    std::vector<int> resultSortedArray = twoSumSortedArray(nums, target);
    if (!resultSortedArray.empty()) {
        std::cout << resultSortedArray[0] << ", " << resultSortedArray[1] << std::endl;
    } else {
        std::cout << "No solution found" << std::endl;
    }

    // Binary Search
    std::cout << "Binary Search: ";
    std::vector<int> resultBinarySearch = twoSumBinarySearch(nums, target);
    if (!resultBinarySearch.empty()) {
        std::cout << resultBinarySearch[0] << ", " << resultBinarySearch[1] << std::endl;
    } else {
        std::cout << "No solution found" << std::endl;
    }

    // Optimized Hash Map
    std::cout << "Optimized Hash Map: ";
    std::vector<int> resultOptimizedHashMap = twoSumOptimizedHashMap(nums, target);
    if (!resultOptimizedHashMap.empty()) {
        std::cout << resultOptimizedHashMap[0] << ", " << resultOptimizedHashMap[1] << std::endl;
    } else {
        std::cout << "No solution found" << std::endl;
    }

    return 0;
}
